Add the partial auth feature of Google account to the Google Ads setup page

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Part of #1000

- Add a new enum `'setup-ads'` for the `next` parameter of GLA's Google auth API.
- Create a new component for asking permission on Google Ads setup page.
- Check required permissions before creating/connecting Google Ads account.
- Add `eventProps` for all tracking events `google_account_connect_button_click` to tell the Google authorization is issued from which page and ask for account authorization or more permission scopes.

### Screenshots:

❗ Please note that these videos were recorded with a local change that the same temporary hack as https://github.com/woocommerce/google-listings-and-ads/pull/1064#discussion_r736426466

![Kapture 2021-10-29 at 14 44 00](https://user-images.githubusercontent.com/17420811/139389201-2b7b381d-f54d-4a79-9842-c6caa7602764.gif)

### Detailed test instructions:

💡  Please ignore the problem of entering Google account choosing page first when going to Google OAuth page. It will be resolved once the dependency PR of #1063 in WCS is merged and deployed.

1. Prepare
    - Disconnect Google and Google Ads accounts from the Connection Test page, and go to the reconnection page.
    - Or, disconnect all accounts and go to the onboarding setup page.
2. (Re)connect Google account without granting the "**Manage your AdWords campaigns**" permission.
3. Go to the Google Ads setup page.
4. It should redirect to Google auth page when clicking on the "Create account" button.
5. After accepting the required permission, should bring you back to the Google Ads setup page.
6. It should display a terms modal when clicking on the "Create account" button, or show existing accounts if your Google account has associated Ads accounts.

### Changelog entry

> Add - The partial auth feature of Google account to the Google Ads setup page.

eason9487 committed:
Create a new component for asking permission on Google Ads setup page

eason9487 committed:
Check required permissions before creating/connecting Google Ads account

eason9487 committed:
Add a new enum "setup-ads" for the "next" parameter of Google auth API.

eason9487 committed:
Add eventProps for tracking event google_account_connect_button_click

ecgan commented on js/src/setup-ads/ads-stepper/setup-accounts/google-ads-account-section/create-account-button.js:
```diff
@@ -10,11 +10,23 @@ import { useState } from '@wordpress/element';
+ *
+ * @param {Object} props React props.
+ * @param {() => boolean|void} [props.onBeforeAskTerms] Called before showing the terms agreement modal.
+ *     Return `true` to interrupt the ask for terms agreement and the subsequent process.
```
❔ From how I see it, it should be "return `false` to interrupt the `handleButtonClick` function and not to ask for terms agreement", or "return `true` to not interrupt the `handleButtonClick` function so it will ask for terms agreement", right? 🤔

eason9487 replied:
Ugh! What was I doing? 😵 
Thanks for catching this! Fixed by https://github.com/woocommerce/google-listings-and-ads/pull/1069/commits/b6d73f3ae3a825495f4033d6337f49282fecab64.

ecgan commented on js/src/setup-ads/ads-stepper/setup-accounts/google-ads-account-section/create-account-button.js:
```diff
@@ -10,11 +10,23 @@ import { useState } from '@wordpress/element';
 	const [ isOpen, setOpen ] = useState( false );
 
 	const handleButtonClick = () => {
+		if ( onBeforeAskTerms && onBeforeAskTerms() === false ) {
```
💅  I always feel this looks easier on the eyes (to cut down the check on `onBeforeAskTerms` at line 27 ) 😄 

```suggestion
	const { onBeforeAskTerms = () => {}, onCreateAccount, ...rest } = props;
	const [ isOpen, setOpen ] = useState( false );

	const handleButtonClick = () => {
		if ( onBeforeAskTerms() === false ) {
```

eason9487 replied:
I added a `noop` function as its default value by https://github.com/woocommerce/google-listings-and-ads/pull/1069/commits/fa9f0a8edee02b786800935115029c9953654468. If want to omit the check, giving a `noop` function could have more semantics and keep the same reference.

ecgan commented on js/src/setup-ads/ads-stepper/setup-accounts/google-ads-account-section/create-account.js:
```diff
@@ -52,6 +51,21 @@ const CreateAccount = ( props ) => {
 	};
 
+	return (
+		<CreateAccountButton
+			loading={ createLoading || fetchAccountLoading }
+			onCreateAccount={ handleCreateAccount }
+		/>
```
❔ Do we want to have the `eventName` and `eventProps` here, like the other `<CreateAccountButton>` element in `AuthorizeAds`? I think we should? 

We didn't have them prior to the code change here, so we can probably do it as a separate issue. 📜

eason9487 replied:
There's already a tracking event `ads_account_create_button_click` in the `<TermsModal>`.

https://github.com/woocommerce/google-listings-and-ads/blob/fa9f0a8edee02b786800935115029c9953654468/js/src/setup-ads/ads-stepper/setup-accounts/google-ads-account-section/terms-modal/index.js#L33-L41

I think it'd be in need if we want to track the drop rate of users leaving the ad account creation after the terms modal showing up.

ecgan commented on js/src/setup-ads/ads-stepper/setup-accounts/google-ads-account-section/authorize-ads.js:
```diff
@@ -0,0 +1,54 @@
+		return false;
+	};
+
+	return (
+		<CreateAccount
+			button={
+				<CreateAccountButton
+					loading={ loading || data }
+					onBeforeAskTerms={ handleBeforeAskTerms }
+					eventName="gla_google_account_connect_button_click"
+					eventProps={ { context: pageName, action: 'scope' } }
+				/>
+			}
+		/>
+	);
```
❓ I have a feeling that we shouldn't use `CreateAccount` and `CreateAccountButton` in this `AuthorizeAds` component here, because the semantic, responsibility and functionality (and also UI, arguably) should be different. Instead, it should be a new component to "request full permission" (not "create account"). Does this make sense?

I'm going to post a separate comment on this with UI design for wider discussion.

ecgan replied:
Posted my comment on UI design below: https://github.com/woocommerce/google-listings-and-ads/pull/1069#issuecomment-954846401

ecgan reviewed (commented):
Tested, working as expected. I left a few comments below. The most important and concerning one (marked with ❓ ) is about the use of `CreateAccount` and `CreateAccountButton` in `AuthorizeAds` component. I'm going to post a separate comment later on this as it requires UI design input. 

Also, we probably need someone to review the `AccountController.php` file. 😄

ecgan commented:
@j111q , referring to the video screenshot gif in the PR description above, when the users have not granted Google Ads permission and they visit the Setup Ads page, they will see this:

![image](https://user-images.githubusercontent.com/417342/139454425-b60d499c-5928-4a24-a669-a5afaac88202.png)

Clicking on the "Create account" button will bring the user to the Google OAuth permission screen, where users have to click the "Allow" button. Then, users will come back to the Setup Ads page, and they will see this, again: 

![image](https://user-images.githubusercontent.com/417342/139454750-b9948794-d9d5-4049-ab30-2650d7639579.png)

Clicking on the "Create account" button here will display the "terms agreement modal", as per existing production behavior.

From the screenshots above, it can be quite confusing to the users, as in "I have already clicked on the Create account button earlier, why am I seeing this again?", and I'm afraid it might end up being logged as a bug.

Here's a suggestion I'm thinking. In the first screenshot above, instead of displaying the "Create your Google Ads account" card, we should have a different card to request for "Manage your AdWords campaigns" permission, similar to the following in Setup MC flow:

![image](https://user-images.githubusercontent.com/417342/139464265-f0e6cc9f-dff4-438d-b325-4ed379e705d8.png)

We should tweak the error text and button text (e.g. "Allow access") so that it is more specific to Google Ads account. What do you think?

@eason9487 , this relates to my comment earlier: https://github.com/woocommerce/google-listings-and-ads/pull/1069#discussion_r739267689. What do you think?

eason9487 committed:
Fix a wrong JSDoc description for the CreateAccountButton component.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1069#discussion_r739248611

eason9487 committed:
Set a noop function as default value of onBeforeAskTerms in the CreateAccountButton component.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1069#discussion_r739252270

eason9487 commented:
@ecgan,

> From the screenshots above, it can be quite confusing to the users, as in "I have already clicked on the Create account button earlier, why am I seeing this again?", and I'm afraid it might end up being logged as a bug.

I think the current design still has its reason. Unlike the onboarding setup, here is an incremental authorization interaction. In many incremental authorization interaction designs, such as camera or GPS access permission, it is common to see the same UI and buttons displayed without first specifically indicating that the user has not yet granted permission. Only when the user takes an "action" that actually requires authorization is the user asked for authorization. This gives the user a reasonable and natural feeling that "because I clicked the open camera button, it makes sense for the app to ask me for permission". And then the user sees the same button again, it's implied but the user should have the awareness that "I'm just done authorizing and the camera button is ready to be used".

However, because GLA must redirect browser to Google authorization page, rather than staying at GLA page and asking for authorization in a pop-up window. This makes the above design approach slightly unnatural. So if this's indeed a concern, then I think it'd be nice to make some adjustments to the button text and color.

findingsimple commented:
> Also, we probably need someone to review the AccountController.php file. 😄

I think it looks ok but perhaps @danielbitzer can 👀

ecgan commented:
@eason9487 ,

> This gives the user a reasonable and natural feeling that "because I clicked the open camera button, it makes sense for the app to ask me for permission". **_And then the user sees the same button again_**, it's implied but the user should have the awareness that "I'm just done authorizing and the camera button is ready to be used".

From my experience in Android mobile apps and web browsers, I don't often see _the same button_. What usually happens for me is: 

1. Click on a Camera button.
2. If there is no access granted, there will be a prompt to grant access. Click allow camera access in the prompt.
3. The app would then have access to the camera, and display the camera view. (We do not see the same camera button, and don't need to click on it again.)

Here's one website that shows what I mean above: https://webcamtests.com/. You would only see and click on the "Test my cam" button once. 

![image](https://user-images.githubusercontent.com/417342/139829019-7b90505e-d55a-4f28-bbca-d0b45b998786.png)

eason9487 commented:
@ecgan, thanks for the explanation! ❤️ 

> From my experience in Android mobile apps and web browsers, I don't often see _the same button_.

Ohhh I got your point. What I meant by "the same button" is the visual of button is no different before/after getting permission when users navigate to a page having that button, no matter which times users visit the page. And your concern is about "continue to proceed with the original action without needing to see and click the same button again".

It could be done by checking whether a URL parameter, which is appended by Google auth page, is existing and then jumping to the terms modal.

Since we will adjust the button text and card description later for the case of lacking permission, users will not see "Create account" then. Instead, after they're brought back to the GLA page to continue, it will be the first time users see the "Create account" button, so I prefer to keep the current implementation of discontinuous actions.

And if we kept the texts as is, your suggestion would bring a better UX for sure. 👍

ecgan commented:
@eason9487 ,

> What I meant by "the same button" is the visual of button is no different before/after getting permission when users navigate to a page having that button, no matter which times users visit the page.

And now I got your point. 😆  👍

findingsimple commented on src/API/Site/Controllers/Google/AccountController.php:
```diff
@@ -25,6 +25,17 @@ class AccountController extends BaseController {
+	 * @var string
+	 */
+	private const NEXT_PATH_MAPPING = [
+		'setup-mc'  => '/google/setup-mc',
```
Consider adding a comment that this first value is used as a default and changing the order of keys/values may affect things below e.g. https://github.com/woocommerce/google-listings-and-ads/pull/1069/files#diff-699dc48ef5ba31467d95145db88b190b3466fc6e935d85117262bbc5cdd72c0bL117

eason9487 replied:
Added by https://github.com/woocommerce/google-listings-and-ads/pull/1069/commits/a4fbf6fc8531bc1439214d1fdf05ff69d82c298e.

findingsimple reviewed (approved):
PHP side of things looks ok to me - nothing glaring, just left a thought to add an additional comment

eason9487 committed:
Add a comment and tweak the @var type for AccountController::NEXT_PATH_MAPPING.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1069#discussion_r745304340

eason9487 commented:
The description and button text/color will have some adjustments with a separate PR based on UX improvements progress (#1065).
